Add partial support for Nordea

noppa opened the pull request:
Hi, and thanks a lot for making this library!

We used this to integrate with Nordea's Web Services API, and modified the library to support that use case.  
If you are interested in upstreaming these changes, here they are :)

I didn't get the certificate requesting part to quite work right with little effort and we don't really need it, with certs generated using other means, so that functionality unfortunately isn't part of this PR. I have included a mention in the readme that the Nordea client lacks the *certificate -methods. 

Other WS API calls work though, and I added a new method `uploadFile` to the client.

I'm happy to answer questions and to try make changes to the implementation if any are needed. As far as I can see, this shouldn't break anything with the Osuuspankki client, but to be honest I don't have a client account there so I haven't really tested either. If possible, I'd recommend testing that it still works before publishing these changes.

Again, thanks for your work

noppa committed:
feat: add partial support for Nordea

This patch refactors OsuuspankkiCertService to its own file (so that we
don't bloat the pankkiyhteys.ts file too much, and then adds classes
similar to OsuuspankkiCertService and Osuuspankki (Client), but for
Nordea.

https://www.nordea.fi/en/business/our-services/mobile-online-services/web-services.html

This does NOT yet implement the certificate request/renewal functions
for Nordea. The user must obtain the Nordea certificate themselves,
for example with the Nordea Security Client that is mentioned in the Web
Services document.

NOTE: Nordea support required adding some new properties to the request
methods. For example, the "Command" parameter is always required by
Nordea. As far as I know this should not break Osuuspankki support, but
that'll have to be verified to avoid regressions.

noppa committed:
fix(types): improve type def soundness

Add type definition for getFileList return value and enforce Nordea's
required file list options by overriding the Client base class file
request methods in Nordea subclass using stricter input params.

Nordea requires options like FileType to be present in file requests.

noppa committed:
feat: add uploadFile method

Add new method to client that allows uploading a file, like a SEPA
payment transfer, using the "uploadFilein" method of the file transfer
service.

noppa committed:
docs: mention Nordea support in the readme

noppa commented:
The build seems to fail due to reduced test coverage. I did play around with the [pankkiyhteys.test.ts](https://github.com/hyrsky/pankkiyhteys/blob/master/test/pankkiyhteys.test.ts) specs and could duplicate some of those Osuuspankki client tests for the Nordea client, but I'm not quite convinced that that would add much real test coverage "safety". The methods tested in those specs mostly fall through to the Client-class, which both of these clients inherit, so that would mostly test the same thing twice. The tests do a lot of mocking and don't (obviously) send request to the actual banks, so they can't really catch errors coming from bank API differences either.

hyrsky commented:
Hi,

Awesome that you found this useful and were able to get Nordea API working! I'll test that my use case with Osuuspankki continues to work with these changes. After that I'll happily merge this. 

Don't worry about coverage limits. I think I'll end up removing it.

noppa commented:
Thanks, sounds good. No rush on our end, we are currently using our version directly from the github fork and that works well enough for now.

I also did some other chore work, upgraded all dependencies and cleaned up some ESLint warnings [here](https://github.com/BoltWorks/pankkiyhteys/compare/feat-nordea-support..upgrade-dependendencies). I didn't include those in this PR because they are kinda irrelevant to the Nordea feature, but I can make another PR if you'd like (perhaps if/when this gets merged).

noppa committed:
docs: minor jsdoc description fix

The uploadFile implementation doesn't actually gzip the file before
sending it. The uploadFilein WS documentation says that compression
would be possible, but I wasn't able to get it working so it's not in
the current implementation. This patch just chnages the
documentation to match the current implementation.
